src/base/dense/owl_base_dense_ndarray_generic.ml

```diff
@@ -2091,6 +2091,129 @@ let avg_pool2d_backward padding input kernel stride output' =
      init_pool_fun add_val_pool_fun end_pool_fun compute_grad_fun)
 
 
+(* TODO: definitely optimise *)
+(* General function for avg_pool3d and max_pool3d *)
+let _pool3d_backward padding input kernel stride output'
+    init_pool_fun add_val_pool_fun end_pool_fun compute_grad_fun =
```

Comment:
Passing higher-order functions then calling them are very expensive operations especially in this scenario because you have so many embedded loops involving a lot of computation. I understand this way of implementing thing can significantly reduce the codebase but also slows down the performance a lot. @tptiplea is currently benchmarking the base ndarray modules, also optimising the functions there. You two @jzstark @tptiplea can probably work together on this.

I will accept current version but it needs to be optimised in future.